README.md
<h1 align="center">
  react-circular-progress-bar
</h1>

<p align="center">
React library to help developers to draw animated, cross-browser, highly customizable progress circles using SVG. IntersectionObserver support, the animation starts when the individual chart appears in the view.
</p>

<p align="center">
  <a href="https://www.npmjs.com/package/@tomik23/react-circular-progress-bar">
    <img src="https://badge.fury.io/js/%40tomik23%2Freact-circular-progress-bar.svg">
  </a>
  <a href="https://bundlephobia.com/package/@tomik23/react-circular-progress-bar">
    <img src="https://img.shields.io/bundlephobia/min/@tomik23/react-circular-progress-bar">
  </a>
  <a href="LICENSE">
    <img src="https://img.shields.io/badge/License-MIT-green.svg">
  </a>
</p>

<p align="center">
  <img src="https://raw.githubusercontent.com/tomik23/react-circular-progress-bar/main/assets/circular-progress-bar.png">
</p>

## Demo

See the demo - [example](https://tomik23.github.io/react-circular-progress-bar/)

## Install

```bash
yarn add @tomik23/react-circular-progress-bar
# or
npm install @tomik23/react-circular-progress-bar
```

## Usage

```jsx
import { CircularProgressBar } from "@tomik23/react-circular-progress-bar";
```

```jsx
// available control variables
const props = {
  percent: 60, // is require
  colorSlice: "#00a1ff",
  colorCircle: "#00a1ff",
  fontColor: "#365b74",
  fontSize: "1.6rem",
  fontWeight: 400,
  size: 200,
  stroke: 10,
  strokeBottom: 5,
  speed: 60,
  cut: 0,
  rotation: -90,
  fill: "#00897B",
  unit: "%",
  textPosition: "0.35em",
  animationOff: false,
  strokeDasharray: "10,1",
  inverse: false,
  round: false,
  number: false,
  linearGradient: ["#ffff00", "brown"],
  styles: {
    borderRadius: "50%",
    boxShadow: "inset 0 0 25px 10px #a2caff"
  }
};

<CircularProgressBar {...props} />;
```

### Update percent

> If you want to update component you have to add `id` to each of them

```jsx
const config = {
  id: 0, // important
  percent: 50,
  colorSlice: "#E91E63"
};

function App() {
  const [update, setUpdate] = useState(config);

  useEffect(() => {
    const interval = setInterval(() => {
      setUpdate({
        ...config,
        id: 0, // we indicate which component we want to change
        percent: Math.floor(Math.random() * 100 + 1)
        colorSlice: "#000",
        fontColor: "#F50057",
        fontSize: "1.2rem",
        fontWeight: 700
      });
    }, 3000);
    return () => clearInterval(interval);
  }, []);

  const newObject = { ...config, ...update };

  return (
    <div>
      <CircularProgressBar {...newObject} />
    </div>
  );
}
```

### Add photos and text

<p align="center">
  <img src="https://raw.githubusercontent.com/tomik23/react-circular-progress-bar/main/assets/image-with-text.png">
</p>

```jsx
const config = {
  percent: 55,
  colorSlice: "#E91E63",
  colorCircle: "#f1f1f1",
  fontWeight: 100,
  number: false // turn off the percentage animation first
};

<CircularProgressBar key={index} {...props}>
  <img
    src="https://picsum.photos/100/100"
    style={{
      width: "60px",
      borderRadius: "50%",
      padding: "2px",
      border: "3px solid salmon"
    }}
    alt="Random image"
  />
  <div style={{ textAlign: "center", padding: "0 35px" }}>
    Lorem, ipsum dolor.
  </div>
</CircularProgressBar>;
```

### Add photos and percent animation

<p align="center">
  <img src="https://raw.githubusercontent.com/tomik23/react-circular-progress-bar/main/assets/image-with-procent.png">
</p>

```jsx
const config = {
  percent: 55,
  colorSlice: "#CDDC39",
  colorCircle: "#f1f1f1",
  fontWeight: 100,
  fontSize: "1rem",
  textPosition: "1.5em" // needed element to move the percentage animation lower
};

<CircularProgressBar key={index} {...props}>
  <img
    src="https://picsum.photos/100/100"
    style={{
      width: "60px",
      borderRadius: "50%",
      marginTop: "-40px",
      padding: "2px",
      border: "3px solid salmon"
    }}
    alt="Random image"
  />
</CircularProgressBar>;
```

### How to turn off `%` or style them?

Turning off the percentage and leaving the number alone is very simple.
Each percent (%) has a class of `circular-unit-x` of course you must add for each circle` id`. If you don't do this, the class will always be `circular-unit-0`. Just add `.circular-unit-x { display: none }` to our styles. Digit animation remains but percent sign [%] disappears.

You can also style this elements to create your own styles, example below:

```css
.circular-unit-15 {
  fill: #f50057;
  font-size: 0.8rem;
  font-weight: 700;
}
```

Instead of adding styles to a component, you can add them in the configuration stage.
```jsx
const examples = [
  {
    percent: 75,
    styles: {
      borderRadius: "50%",
      boxShadow: "inset 0 0 25px 10px #a2caff"
    }
  }
];

{examples.map((props, index) => {
  return <CircularSection key={index} {...props} />;
})}
```
See an example [Minimal Width Shadow](https://tomik23.github.io/react-circular-progress-bar/?path=/story/circular-progress-bar--minimal-width-shadow)

## Configuration of the plugin

| props           |  type   |   default   | require | description |
| --------------- | :-----: | :---------: | :-----: | ----- |
| percent         | number  |             |    ✔    | Represents the progress bar and animation of the animation progress expressed by a number e.g. 65% |
| id              | number  |             |         | If you want to update a component, you need to add an `id` to each of them. Also when you want to display several components with different gradients - `linearGradient` |
| speed           | number  |             |         | Frame rate animation [fps]. Let's say you want the animation to be 60fps, just add the parameter `speed: 60` |
| animationOff    | boolean |   `false`   |         | Turn off the progress animation |
| animationSmooth | string  |   ``        |         | The [transition](https://developer.mozilla.org/en-US/docs/Web/CSS/transition) property allows you to create animations (a smooth transition effect), e.g. `500ms ease-out` |
| colorSlice      | string  | `'#00a1ff'` |         | Progress layer color and background ["#ffff00","brown" \*](#colors-names) |
| colorCircle     | string  | `'#00a1ff'` |         | Bottom circle color Font ["#ffff00","brown" \*](#colors-names) |
| stroke          | number  |    `10`     |         | Stroke width, chart thickness |
| strokeBottom    | number  |    `10`     |         | If "strokBottom" is set, it is used to generate a background circle size |
| round           | boolean |   `false`   |         | Path rounding |
| inverse         | boolean |   `false`   |         | Counterclockwise animation |
| rotation        | number  |    `-90`    |         | Chart rotation |
| number          | boolean |   `true`    |         | Add props number and set to false to hide the number with percent |
| size            | number  |    `200`    |         | Size progress bar width and height in px |
| cut             | number  |     `0`     |         | Angle of the circle sector |
| unit            | string  |     `%`     |         | Different unit instead of percentage (%) inside the circle |
| fill            | string  |   `none`    |         | Inner circle background color |
| strokeDasharray | string  |     ``      |         | It works only on the lowest circle and only on whole circles - [stroke-dasharray](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/) |
| textPosition    | string  |  `0.35em`   |         | The position of the SVG TEXT element vertically |
| fontSize        | string  |  `1.6rem`   |         | Font size. The font can be shown in units rem, em, px ... |
| fontWeight      | number  |    `400`    |         | 400, 600, ... |
| fontColor       | string  | `'#365b74'` |         | Font color ["#ffff00","brown" \*](#colors-names) |
| linearGradient  |  array  |             |         | Array of colors "lineargradient": ["#ffff00","brown" \*](#colors-names) |
| styles          |  object |             |         | Possibility to add styles to parent div see example [Minimal Width Shadow](https://tomik23.github.io/react-circular-progress-bar/?path=/story/circular-progress-bar--minimal-width-shadow) |

## Colors names

[\* See colors names](https://htmlcolorcodes.com/color-names/)

## Browser support

react-circular-progress-bar supports all major browsers including IE 11 and above

CHANGELOG.md
## v1.1.1 (2022-01-06)

### Added
- `styles` - new props
- an example of using the `styles` prop
### Changed
- remove extra `styleObj` div

## v1.1.0 (2022-01-03)

### Changed
- updating accessibility, change role from `img` to `progressbar`
- adding `aria-valuenow`, `aria-valuemin` and `aria-valuemax` attributes

### Fixed
- test width `strokeDasharray`

## v1.0.25 (2021-10-21)

### Added
- `animationSmooth` - new props, The [transition](https://developer.mozilla.org/en-US/docs/Web/CSS/transition) property allows you to create animations (a smooth transition effect). 
### Changed

- `role=img` to `role=progressbar` - [progressbar role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role)

## v1.0.24 (2021-09-27)

### Changed

- rebuilded presentation examples 

## v1.0.23 (2021-09-24)

### Added

- 'strokeDasharray' - added to the lower circle
- more examples to storybook

## v0.1.22 (2021-08-23)

### Changed

- a separate element for a digits (tspan - `circular-percent-x`)
- the color of the `text` element to `#000`
- updated style storybook `unit`
- upgrating storybook to 6.3.7

## v0.1.21 (2021-08-20)

### Added

- `unit` - different unit instead of percentage (%) inside the circle
- `fill` - inner circle color, defaults `fill: "none"`

## v0.1.19 (2021-08-19)

### Added

- `cut` - angle of the circle sector
- `rotation` - chart rotation


src/App.js
import React, { useState, useEffect, Fragment } from "react";
import CircularProgressBar from "./components/CircularProgressBar";

const examples = [
  {
    id: 0,
    name: "minimal",
    percent: 75
  },
  {
    id: 1,
    name: "minimal width shadow",
    percent: 75,
    styles: {
      borderRadius: "50%",
      boxShadow: "inset 0 0 25px 10px #a2caff"
    }
  },
  {
    id: 2,
    name: "simple",
    colorSlice: "#BF360C",
    colorCircle: "#f1f1f1",
    fontWeight: 100,
    percent: 75
  },
  {
    id: 3,
    name: "linear gradient",
    linearGradient: ["yellow", "#ff0000"],
    percent: 75,
    colorCircle: "#e6e6e6",
    round: true
  },
  {
    id: 4,
    name: "counterclockwise",
    inverse: true,
    percent: 75,
    colorSlice: "#AB47BC",
    colorCircle: "#f1f1f1",
    fontWeight: 100
  },
  {
    id: 5,
    name: "stroke-dasharray",
    strokeDasharray: "1, 2",
    percent: 75,
    colorSlice: "#F5F5F5",
    colorCircle: "#424242",
    stroke: 5,
    strokeBottom: 10,
    round: true
  },
  {
    id: 6,
    name: "animation off",
    animationOff: true,
    percent: 75,
    colorSlice: "#CDDC39",
    colorCircle: "#f1f1f1",
    fontWeight: 100
  },
  {
    id: 7,
    name: "without number",
    number: false,
    percent: 75,
    colorSlice: "#000",
    colorCircle: "#e6e6e6"
  },
  {
    id: 8,
    name: "rotation",
    rotation: 90,
    percent: 75,
    colorSlice: "#ff8c69",
    colorCircle: "#f1f1f1",
    fontWeight: 100
  },
  {
    id: 9,
    name: "fill",
    fill: "#EFEBE9",
    percent: 75,
    stroke: 6,
    strokeBottom: 10,
    colorSlice: "#4E342E",
    colorCircle: "#EFEBE9",
    fontSize: "1.3em",
    round: true
  },
  {
    id: 10,
    name: "unit",
    unit: "PL",
    percent: 75,
    stroke: 5,
    strokeBottom: 10,
    colorSlice: "#fff",
    colorCircle: "#F50057",
    round: true
  },
  {
    id: 11,
    name: "stroke-bottom",
    strokeBottom: 5,
    percent: 75,
    colorSlice: "#EC407A",
    colorCircle: "#f1f1f1",
    round: true
  },
  {
    id: 12,
    name: "speed",
    speed: 30,
    percent: 75,
    colorSlice: "#DD2C00",
    colorCircle: "#f1f1f1",
    round: true
  },
  {
    id: 13,
    name: "cut",
    cut: 30,
    percent: 75,
    rotation: 144,
    colorSlice: "#F50057",
    colorCircle: "#f1f1f1",
    textPosition: "1em",
    fontSize: "1.2rem"
  },
  {
    id: 14,
    name: "textposition",
    percent: 75,
    textPosition: "1.5em",
    colorSlice: "#880E4F",
    colorCircle: "#f1f1f1",
    fontSize: "1em"
  },
  {
    id: 15,
    name: "mixed",
    percent: 75,
    textPosition: "0em",
    colorSlice: "#FF1744",
    colorCircle: "#fff",
    stroke: 5,
    strokeBottom: 10,
    strokeDasharray: "1, 3",
    fontSize: "1.2rem",
    fontWeight: 600,
    round: true,
    styles: {
      borderRadius: "50%",
      boxShadow: "inset 0 0 25px 10px #f50057"
    }
  },
  {
    id: 16,
    name: "smooth animation",
    animationSmooth: "1s ease-out",
    percent: 75,
    colorSlice: "#FF6D00",
    colorCircle: "#f1f1f1"
  }
];

const update = [
  {
    id: 55,
    name: "update",
    percent: 75,
    colorSlice: "#BF360C",
    colorCircle: "#f1f1f1",
    fontWeight: 100
  }
];

function CircularSection(object) {
  const [isActive, setIsActive] = useState(false);

  // remove name from object for presentation
  const newObject = Object.keys(object).reduce((obj, key) => {
    if (key !== "name") {
      obj[key] = object[key];
    }
    return obj;
  }, {});

  return (
    <section>
      <div className="flex info">
        <div className="info-name">{object.name}</div>
        <div className="info-code" onClick={() => setIsActive(!isActive)}>
          code
        </div>
      </div>
      <CircularProgressBar {...object}>
        {object.id === 14 && (
          <img
            src="https://picsum.photos/100/100"
            style={{
              position: "absolute",
              width: "50px",
              borderRadius: "50%",
              left: "50%",
              top: "50%",
              transform: "translate(-50%, -90%)",
              padding: "2px",
              border: "3px solid salmon"
            }}
            alt="Random"
          />
        )}
        {object.id === 15 && (
          <Fragment>
            <div
              style={{
                position: "absolute",
                fontSize: "2rem",
                margin: "auto",
                width: "100%",
                bottom: "45px",
                textAlign: "center",
                animation: "heart 0.8s linear infinite",
                textShadow: "0px 0px 10px #f50057"
              }}>
              &#10084;
            </div>
          </Fragment>
        )}
      </CircularProgressBar>
      {isActive && (
        <div className="precode">
          <pre>
            <code>{JSON.stringify(newObject, null, 2)}</code>
          </pre>
        </div>
      )}
    </section>
  );
}

function App() {
  const [progress, setProgress] = useState(75);
  const percent = { percent: progress };

  const [first, setFirst] = useState(update[0]);
  useEffect(() => {
    const interval = setInterval(() => {
      const typeFont = [100, 200, 300, 400, 500, 600, 700];
      const colorHex = `#${Math.floor((Math.random() * 0xffffff) << 0).toString(
        16
      )}`;
      setFirst({
        ...update[0],
        id: 55,
        percent: Math.floor(Math.random() * 100 + 1),
        colorSlice: colorHex,
        fontColor: colorHex,
        fontSize: `${Math.floor(Math.random() * (1.3 - 1 + 1) + 1)}rem`,
        fontWeight: typeFont[Math.floor(Math.random() * typeFont.length)]
      });
    }, 3000);
    return () => clearInterval(interval);
  }, []);

  return (
    <Fragment>
      <h1>circular-progress-bar</h1>
      <div className="flex progress-wrapper">
        <div className="progress">progress</div>
        <input
          type="range"
          value={progress}
          min="0"
          max="100"
          onChange={(event) => setProgress(parseInt(event.target.value, 10))}
        />
      </div>
      <div className="flex container">
        {examples.map((props, index) => {
          const obj = { ...props, ...percent };
          return <CircularSection key={index} {...obj} />;
        })}

        {update.map((props, index) => {
          const obj = { ...props, ...percent, ...first };
          return <CircularSection key={index} {...obj} />;
        })}
      </div>
    </Fragment>
  );
}

export default App;
